src/supervisor/kuberenetes-api-wrappers.ts

```diff
@@ -1,32 +1,46 @@
 import * as http from 'http';
 import * as sleep from 'sleep-promise';
+import { IRequestError } from './types';
 
 export const ATTEMPTS_MAX = 3;
 export const DEFAULT_SLEEP_SEC = 1;
 export const MAX_SLEEP_SEC = 5;
+const HTTP_STATUS_TOO_MANY_REQUESTS = 429;
 type IKubernetesApiFunction<ResponseType> = () => Promise<ResponseType>;
 
+function getErrorCode(err: IRequestError): number | undefined {
+  if (!(err.response)) {
+    return undefined;
+  }
+
+  return  err.response.statusCode;
+}
+
+function shouldRetryRequest(err: IRequestError, attempt: number): boolean {
+  if (attempt === ATTEMPTS_MAX) {
+    return false;
+  }
+
+  const code = getErrorCode(err);
+  if (code === HTTP_STATUS_TOO_MANY_REQUESTS) {
+    return true;
+  }
+
+  return false;
+}
+
 export async function retryKubernetesApiRequest<ResponseType>(
   func: IKubernetesApiFunction<ResponseType>
 ): Promise<ResponseType> {
   for (let attempt = 1; attempt <= ATTEMPTS_MAX; attempt++) {
     try {
       return await func();
     } catch (err) {
-      if (!(err.response)) {
-        throw err;
-      }
```

Comment:
This logic has now disappeared, which means that this new code on line 43:
```js
const sleepSeconds = calculateSleepSeconds(err.response);
```
Can now pass an undefined response to `calculateSleepSeconds` and cause an error!
So the logic of the function has changed, is it safe to do so?

If we keep this piece of code, then we don't have to return `undefined` in `getErrorCode()`. In fact, we also probably don't need a function there, since we can just directly reference `response.statusCode`!

Reply:
I was putting it in a way that it could help me to include new errors. But I think it would be good go a step back and rethink.

Reply:
I think that keeping that piece of code is more accurate, because now the code on line 37 (calculateSleepSeconds(...)) relies on some previous flow checking that the response is present, which is encapsulated in a function. If some day that check is removed, we'll probably break calculateSleepSeconds().

I think the existing flow is much more accurate:
- ensure the response is not falsy
- do something with the response
- do something else with the response

Otherwise, it becomes:
- do something with the response, but only if it's not empty
- do something else with the response (potentially also if it's not empty?)

I think we could catch these cases by changing the type to help us:
```js
export interface IRequestError {
  response?: IncomingMessage; // see how response is now optional
}
```

Maybe this would help surface the issue I'm trying to guard against?

In any case, it looks like we do handle an undefined response already: https://github.com/snyk/kubernetes-monitor/blob/89134966363adb3e160dcf84be206afc9223988b/src/supervisor/kuberenetes-api-wrappers.ts#L39
 so my suggestion would apply to any future places where we write similar code, to keep in mind not to rely too much on previous checks and handle the undefined check only in one place, at the top level 🙏